src/util/thread_pool.hpp

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+#include "functional.hpp"
+
+namespace taraxa::util {
+
+class ThreadPool : std::enable_shared_from_this<ThreadPool> {
+  boost::asio::io_context ioc_;
+  boost::asio::executor_work_guard<decltype(ioc_)::executor_type> ioc_work_;
+  std::vector<std::thread> threads_;
+  std::mutex threads_mu_;
+
+ public:
+  explicit ThreadPool(size_t num_threads, bool _start = true);
+  ~ThreadPool();
+
+  ThreadPool(ThreadPool const &) = delete;
+  ThreadPool &operator=(ThreadPool const &) = delete;
+
+  auto &unsafe_get_io_context() { return ioc_; }
+
```

Comment:
I Dont think this is a good design. I went through code how you use this threadpool and it is widely used just by accessing io_service, which is retrieved by reference from this method... Then this io_context is sent to the generic boost::asio::post methods and it is very unclear that tasks are actually going to be processed by this threadpool. I had to track down where is that io_service from and found out that  you sent reference to io_context to the taraxa)capability constructor and then use post methos from util/boost_asio.hpp... It is just too complicated. Why not just sent this threadpool to the taraxa-capability and use post methods on it ? io_context should be totally encapsulated inside with no option to access it but class methods

Reply:
This is a tip of an iceberg. I copied this code from the branch I'm working on in parallel, and there `unsafe_get_io_context` is justified, because I use the thread pool class as a unified scheduler, and unfortunately some our dependencies can only work with `io_context` interface. I called it `unsafe....` precisely to suggest that it is bad but we have to live with it.